lib/user_comm_screener.rb

```diff
@@ -0,0 +1,208 @@
+# action (such as raising an error or showing a helpful message) if so.
+#
+# Users may Mute another user (the actor), which will:
+#
+#   * Prevent PMs from the actor
+#   * Prevent notifications from the actor
+#
+# Users may Ignore another user (the actor), which will:
+#
+#   * Do everything that Mute does as well as suppressing content made by
+#     the actor (such as posts) from the UI
+#
+# Users may also either:
+#
+#   a) disallow all PMs sent to them or
+#   b) disallow PMs except from a certain allowlist of users
+#
+# A user may have this preference but have no Muted or Ignored users, which
+# necessitates the difference between methods in this class.
+#
+# An important note is that **all of these settings do not apply when the actor
+# is a staff member**. So admins and moderators can PM and notify anyone they please.
+class UserCommScreener
+  attr_reader :acting_user, :preferences
+
+  class UserCommPref
+    attr_accessor :username, :is_muting, :is_ignoring, :is_disallowing_all_pms,
+      :is_disallowing_pms_from_acting_user
+
+    def initialize(preferences)
+      @username = preferences[:username]
+      @is_muting = preferences[:is_muting]
+      @is_ignoring = preferences[:is_ignoring]
+      @is_disallowing_all_pms = preferences[:is_disallowing_all_pms]
+      @is_disallowing_pms_from_acting_user = preferences[:is_disallowing_pms_from_acting_user]
+    end
+
+    def communication_prevented?
+      ignoring_or_muting? || disallowing_pms?
+    end
+
+    def ignoring_or_muting?
+      is_muting || is_ignoring
+    end
+
+    def disallowing_pms?
+      is_disallowing_all_pms || is_disallowing_pms_from_acting_user
+    end
+  end
+
+  UserCommPrefs = Struct.new(:acting_user, :user_preference_map) do
+    def acting_user_staff?
+      acting_user.staff?
+    end
+
+    def usernames
+      user_preference_map.values.map(&:username)
+    end
+
+    def for_user(user_id)
+      user_preference_map[user_id]
+    end
+
+    def allowing_actor_communication
+      return user_preference_map.values if acting_user_staff?
+      user_preference_map.reject do |user_id, pref|
+        pref.communication_prevented?
+      end.values
+    end
+
+    def preventing_actor_communication
+      return [] if acting_user_staff?
+      user_preference_map.select do |user_id, pref|
+        pref.communication_prevented?
+      end.values
+    end
+
+    def ignoring_or_muting?(user_id)
+      return false if acting_user_staff?
+      pref = for_user(user_id)
+      pref.present? && pref.ignoring_or_muting?
+    end
+
+    def disallowing_pms?(user_id)
+      return false if acting_user_staff?
+      pref = for_user(user_id)
+      pref.present? && pref.disallowing_pms?
+    end
+  end
+  private_constant :UserCommPref
+  private_constant :UserCommPrefs
+
+  def initialize(acting_user, target_usernames:)
+    @acting_user = acting_user.is_a?(Integer) ? User.find(acting_user) : acting_user
```

Comment:
I would prefer adding `acting_user_id` as an optional keyword argument and raise an error if `acting_user` and `acting_user_id` are both used when initialising an object. I've always felt that having a variable accept either an id or object is very confusing.